libraries/AP_Scripting/applets/Script_Controller.lua

```diff
@@ -0,0 +1,185 @@
+   local block_size = 256
+   local file1 = io.open(filename1, "rb")
+   if not file1 then
+      return false
+   end
+   local file2 = io.open(filename2, "rb")
+   if not file2 then
+      file1:close()
+      return false
+   end
+   local ret = true
+   while true do
+      local block1 = file1:read(block_size)
+      local block2 = file2:read(block_size)
+      if block1 ~= block2 then
+         ret = false
+         break
+      end
+      if not block1 then
+         break
+      end
+   end
+   file1:close()
+   file2:close()
+   return ret
+end
+
+--[[
+   get the path to the scripts directory. This will be scripts/ on SITL
+   and APM/scripts on a ChibiOS board
+--]]
+function get_scripts_dir()
+   local dlist1 = dirlist("APM/scripts")
+   if dlist1 and #dlist1 > 0 then
+      return "APM/scripts"
+   end
+   -- otherwise assume scripts/
+   return "scripts"
+end
+
+--[[
+    check if a file exists, returns true if it does' fname is complete path to file
+--]]
+function file_exists(fname)
+   local f = io.open(fname,"rb")
+   if not f then
+      return false
+   end
+   f:close()
+   return true
+end
+
+--[[
+   remove any lua scripts in the scripts directory that are not in the given subdir besides ourselves
+   returns true if any files were removed
+--]]
+function remove_scripts(subdir)
+   local sdir = get_scripts_dir()
+   local dlist = dirlist(sdir)
+   if not dlist then
+      return false
+   end
+   local ret = false
+   for k,v in ipairs(dlist) do
+      local suffix = v:sub(-4)
+      if suffix == ".lua" and v ~= THIS_SCRIPT then
+         if not file_exists(subdir .. "/" .. v) then
+            ret = true
+            remove(sdir .. "/" .. v)
+         end
+      end
+   end
+   return ret
+end
+
+--[[
+   copy scripts from a subdir to the main scripts directory
+   return true if any changes were made
+--]]
+function copy_scripts(subdir)
+   local dlist = dirlist(subdir)
+   if not dlist then
+      return false
+   end
+   local ret = false
+   local sdir = get_scripts_dir()
+   for k, v in ipairs(dlist) do
+      local suffix = v:sub(-4)
+      if suffix == ".lua" and v ~= THIS_SCRIPT then
+         local src = subdir .. "/" .. v
+         local dest = sdir .. "/" .. v
+         if not file_compare(src, dest) then
+            ret = true
+            file_copy(src, dest)
+         end
+      end
+   end
+   if ret then
+      gcs:send_text(5,"Copied new files to scripts directory")
+   end
+   return ret
+end
+
+--[[
+   activate a scripting subdirectory
+--]]
+function activate_subdir(n)
+   -- step1, remove lua files from scripts/ that are not in the given subdirectory
+   local subdir = get_scripts_dir() .. "/" .. n
+   local changes_made = remove_scripts(subdir)
+   changes_made = changes_made or copy_scripts(subdir)  --copy files if different
+   return changes_made
+end
+
+local sw_last = 0
+function update()
+   local sw_current = scripting_dir_select:norm_input_ignore_trim() + 2
```

Comment:
this needs to use the get_aux_cached() method
see aux_cached.lua example

Reply:
done